heclib/heclib_c/src/Public/zpdRetrieve.c

```diff
@@ -705,8 +706,272 @@ int zpdRetrieve(long long *ifltab, zStructPairedData *pds, int retrieveSizeFlag)
+		if (cp && !strncasecmp(cp, "ELEV", 4)) {
+			depElev = TRUE;
+		}
+		if (indElev || depElev) {
+			//------------------------//
+			// elevation in parameter //
+			//------------------------//
+			zquery("VDTM", cvertical_datum, sizeof(cvertical_datum), &ivertical_datum);
+			char cvertical_datum[CVERTICAL_DATUM_SIZE];
+			int  ivertical_datum = -1;
+			verticalDatumInfo _vdi;
+			verticalDatumInfo *vdi = NULL;
+			char *vdiStr;
+			char errmsg[1024];
+			zquery("VDTM", cvertical_datum, sizeof(cvertical_datum), &ivertical_datum);
+			if (ivertical_datum != IVERTICAL_DATUM_UNSET) {
+				//-----------------------------------//
+				// specific vertical datum requested //
+				//-----------------------------------//
+				if (pds->locationStruct && pds->locationStruct->supplemental) {
+					vdiStr = extractFromDelimitedString(
+						&pds->locationStruct->supplemental,
+						VERTICAL_DATUM_INFO_USER_HEADER_PARAM,
+						":",
+						TRUE,
+						FALSE,
+						'\n');
+					if (vdiStr) {
+						char *msg = stringToVerticalDatumInfo(&_vdi, vdiStr);
+						if (msg) {
+							char _msg[65];
+							strncpy(_msg, msg, sizeof(_msg)-1);
+							sprintf(
+								errmsg,
+								"Cannot convert from native vertical datum of '%s' to '%s'.\n%s\n"
+								"Values not converted.",
+								pds->locationStruct->verticalDatum == 0 ? "UNSET"   :
+								pds->locationStruct->verticalDatum == 1 ? "NAVD-88" :
+								pds->locationStruct->verticalDatum == 2 ? "NGVD-29" : "OTHER",
+								cvertical_datum,
+								_msg);
+							zmessage(ifltab, errmsg);
+						}
+						else {
+							vdi = &_vdi;
+						}
+						free(vdiStr);
+					}
+				}
+				if (vdi == NULL) {
+					sprintf(
+						errmsg,
+						"Cannot convert from native vertical datum of '%s' to '%s'.\n"
+						"Record has no conversion information.\n"
+						"Values not converted.",
+						pds->locationStruct->verticalDatum == 0 ? "UNSET"   :
+						pds->locationStruct->verticalDatum == 1 ? "NAVD-88" :
+						pds->locationStruct->verticalDatum == 2 ? "NGVD-29" : "OTHER",
+						cvertical_datum);
+					zmessage(ifltab, errmsg);
+				}
+				else {
+					//---------------------------------------------------------------//
+					// ensure the vertical datum info is returned in the user header //
+					//---------------------------------------------------------------//
+					verticalDatumInfoToString(&vdiStr, vdi, TRUE);
+					char *headerString;
+					if (pds->userHeader) {
+						headerString = userHeaderToString(pds->userHeader, pds->userHeaderNumber);
+						if (!strstr(headerString, VERTICAL_DATUM_INFO_USER_HEADER_PARAM)) {
+							headerString = (char *)realloc(
+								headerString,
+								strlen(headerString)
+								+ VERTICAL_DATUM_INFO_USER_HEADER_PARAM_LEN
+								+ strlen(vdiStr)
+								+ 3);
+							sprintf(
+								headerString+strlen(headerString), 
+								"%s:%s;", 
+								VERTICAL_DATUM_INFO_USER_HEADER_PARAM,
+								vdiStr);
+							free(pds->userHeader);
+						}
```

Comment:
Should handle the case where VDI is already in user header. It shouldn't happen, but if it does, the location VDI should take precedence.